### PR TITLE
Update the metric learning vision example to use Tensorflow Similarity.

### PR DESCRIPTION
**Pull request by owenvallis**

### CODE REVIEW

File: examples/vision/metric_learning.py [hunk 1]
```diff
@@ -22,290 +24,380 @@
...
+
+The following cell creates a train_ds sample that:
+
+* Loads the cifar10 dataset from tfds and then takes the
+examples_per_class_per_batch.
+* Normalizes the images by applying the preprocess function to each image.
+* Ensures the sampler restricts the classes to those defined in the class_list.
+* Ensures each batch contains 4 different classes with 16 examples each.
+
+We also create a train_ds in the same way, but we limit the total number of
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:93]: Why do you need two train ?

File: examples/vision/metric_learning.py [hunk 2]
```diff
@@ -22,290 +24,380 @@
...
+per class. Additionally, we can restrict the sampler to only yield the subset of
+classes defined in class_list, enabling us to train on a subset of the classes
+and then test how the embedding generalizes to the unseen classes. This can be
+useful when working on few-shot learning problems.
+
+The following cell creates a train_ds sample that:
+
+* Loads the cifar10 dataset from tfds and then takes the
+examples_per_class_per_batch.
+* Normalizes the images by applying the preprocess function to each image.
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:89]: I rather have the rescaling as part of model its more keras.

File: examples/vision/metric_learning.py [hunk 3]
```diff
@@ -22,290 +24,380 @@
...
+* Ensures the sampler restricts the classes to those defined in the class_list.
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:90]: Are you restricting to N class for training -- if so make it explicit.

File: examples/vision/metric_learning.py [hunk 4]
```diff
@@ -22,290 +24,380 @@
...
+* Ensures each batch contains 4 different classes with 16 examples each.
+
+We also create a train_ds in the same way, but we limit the total number of
+examples per class to 100 and the examples per class per batch is set to the
+default of 2.
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:95]: This is not what the code do below. Please make sure its consistent.

File: examples/vision/metric_learning.py [hunk 5]
```diff
@@ -22,290 +24,380 @@
...
-        size=(num_cols * box_size, num_rows * box_size),
-        color=(250, 250, 250),
-    )
-    for row_idx in range(num_rows):
-        for col_idx in range(num_cols):
-            array = (np.array(examples[row_idx, col_idx]) * 255).astype(np.uint8)
-            collage.paste(
-                Image.fromarray(array), (col_idx * box_size, row_idx * box_size)
-            )
+examples_per_class_per_batch = 8
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:105]: Allows user to explicitly select the number of class they want and don't randomize the order -- its introduce unnecessary variance. class_per_batch == distinct_class unify this too.

File: examples/vision/metric_learning.py [hunk 6]
```diff
@@ -22,290 +24,380 @@
...
-    # Double size for visualisation.
-    collage = collage.resize((2 * num_cols * box_size, 2 * num_rows * box_size))
-    return collage
+print(" Create Training Data ".center(34, '#'))
+train_ds = TFDatasetMultiShotMemorySampler(
+    'cifar10',
+    classes_per_batch=classes_per_batch,
+    splits='train',
+    steps_per_epoch=5000,
+    examples_per_class_per_batch=examples_per_class_per_batch,
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:115]: discuss why you want 8 or even more per class --> I think if you are using a TPU you need way more to saturate it.

File: examples/vision/metric_learning.py [hunk 7]
```diff
@@ -22,290 +24,380 @@
...
+    ax.imshow(im)
+    ax.axis('off')
 
 """
-For this example we are using the simplest approach to training; a batch will consist of
-`(anchor, positive)` pairs spread across the classes. The goal of learning will be to
-move the anchor and positive pairs closer together and further away from other instances
-in the batch. In this case the batch size will be dictated by the number of classes; for
-CIFAR-10 this is 10.
+## Batch Sizes
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:148]: We don't need this -- its a distraction from the colab intent -- instead add a sentence when you talk to get slice.

File: examples/vision/metric_learning.py [hunk 8]
```diff
@@ -22,290 +24,380 @@
...
-            positive_idx = random.choice(examples_for_class)
-            while positive_idx == anchor_idx:
-                positive_idx = random.choice(examples_for_class)
-            x[0, class_idx] = x_train[anchor_idx]
-            x[1, class_idx] = x_train[positive_idx]
-        return x
+* indexing embedded examples
+* performing example lookups
+* evaluating the classification
+* evaluating the quality of the embedding space
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:173]: point to the git repo.

File: examples/vision/metric_learning.py [hunk 9]
```diff
@@ -22,290 +24,380 @@
 
 """
 ## Setup
+
+This tutorial will use [Tensorflow Similarity](https://github.com/tensorflow/similarity)
+to learn and evaluate the similarity embedding. The tensorflow similarity
+package provides a number of components that:
+
+* Simplify constructing contrastive models.
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:32]: Makes training contrastive model simple and fast?

File: examples/vision/metric_learning.py [hunk 10]
```diff
@@ -22,290 +24,380 @@
...
+* performing example lookups
+* evaluating the classification
+* evaluating the quality of the embedding space
+"""
 
+embedding_size = 256
+
+inputs = layers.Input((32, 32, 3))
+x = inputs
+x = layers.Conv2D(64, 3, activation='relu')(x)
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:180]: Add the normalization and a simple aug layers.

File: examples/vision/metric_learning.py [hunk 11]
```diff
@@ -22,290 +24,380 @@
...
 
 """
-We can visualise a batch in another collage. The top row shows randomly chosen anchors
-from the 10 classes, the bottom row shows the corresponding 10 positives.
+## Similarity Loss
+
+The similarity loss expects batches containing at least 2 examples of each
+class, from which it computes the loss over the pairwise positive and negative
+distances. Here we are using `MultiSimilarityLoss()`, one of several losses in
+Tensorflow Similarity. This loss attempts to use all informative pairs in the
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:201]: Linkify TF similarity

File: examples/vision/metric_learning.py [hunk 12]
```diff
@@ -22,290 +24,380 @@
...
+batch, taking into account the self-similarity, positive-similarity, and the
+negative-similarity.
+
+For more information on the MS loss see
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:205]: Remove and simply link the paper next to the MultiSimilariyt() part.

File: examples/vision/metric_learning.py [hunk 13]
```diff
@@ -22,290 +24,380 @@
...
-gram_matrix = np.einsum("ae,be->ab", embeddings, embeddings)
-near_neighbours = np.argsort(gram_matrix.T)[:, -(near_neighbours_per_example + 1) :]
+# This yields 100 examples for each class.
+# We defined this when we created the test_ds sampler.
+x_confusion, y_confusion = test_ds.get_slice(0, -1)
+
+matches = model.match(x_confusion, cutpoint=cutpoint, no_match_label=10)
+confusion_matrix(matches,
+                 y_confusion,
+                 labels=labels,
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:359]: Label should be the classes names, I don't think its the case? They are listed below.

File: examples/vision/metric_learning.py [hunk 14]
```diff
@@ -22,290 +24,380 @@
...
-    "Dog",
-    "Frog",
-    "Horse",
-    "Ship",
-    "Truck",
-]
-disp = ConfusionMatrixDisplay(confusion_matrix=confusion_matrix, display_labels=labels)
-disp.plot(include_values=True, cmap="viridis", ax=None, xticks_rotation="vertical")
-plt.show()
+num_examples_to_clusters = 720
```
Comment (ebursztein) [inline examples/vision/metric_learning.py:391]: 720 is a weird number there are 10 classes so 700?

**Review by ebursztein — CHANGES_REQUESTED** [review]

File: examples/vision/metric_learning.py (hunk 1 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:93]: should be test_ds, fixed

File: examples/vision/metric_learning.py (hunk 3 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:90]: We are using all 10 classes, but I updated line # 105 to make it clearer that the user can change this to restrict the number of classes used during training.

File: examples/vision/metric_learning.py (hunk 4 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:95]: This was meant to refer to the test_ds set. I'll add a similar section to describe the train_ds params.

File: examples/vision/metric_learning.py (hunk 7 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:148]: Done

File: examples/vision/metric_learning.py [hunk 15]
```diff
@@ -1,18 +1,20 @@
 """
 Title: Metric learning for image similarity search
-Author: [Mat Kelcey](https://twitter.com/mat_kelcey)
-Date created: 2020/06/05
-Last modified: 2020/06/09
+Author: [Owen Vallis](https://twitter.com/owenvallis)
```
Comment (fchollet) [inline examples/vision/metric_learning.py:3]: This appears to be a very extensive rewrite of the example. Prefer creating a new example besides the old one rather than overwriting the old one. Only if the changes are relatively limited should we update an existing example.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR. This looks like an extensive rewrite, so please add a new example rather than overwriting the current one. Please only add the Python file to the PR. The generated files should only be added once the code and copy are final and the PR has been approved.

File: examples/vision/metric_learning.py (hunk 8 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:173]: Done

File: examples/vision/metric_learning.py (hunk 9 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:32]: Done

File: examples/vision/metric_learning.py (hunk 11 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:201]: Done

File: examples/vision/metric_learning.py (hunk 12 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:205]: Done

File: examples/vision/metric_learning.py (hunk 14 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:391]: Done

File: examples/vision/metric_learning.py (hunk 15 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:3]: Sounds good, I'll rename it and add the original example back. Given that the two examples are very similar, should I name this one so that it's clear they are related, maybe something like `metric_learning_tf_similarity.py`?

File: examples/vision/metric_learning.py (hunk 13 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:359]: I updated this as part of https://github.com/tensorflow/similarity/pull/170.

`labels` should now be a list of strs and `class_mapping` should now be the map from class ids to class labels.

File: examples/vision/metric_learning.py (hunk 10 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:180]: Added the normalization but we ended up omitting the augmentation layers here as they didn't seem to improve the loss here.

File: examples/vision/metric_learning.py (hunk 15 above)
Comment (fchollet) [reply examples/vision/metric_learning.py:3]: Sounds good! Definitely do point out "TF Similarity" in the title as the differentiating aspect of this example.

File: examples/vision/metric_learning.py (hunk 6 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:115]: Done

File: examples/vision/metric_learning.py (hunk 2 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py:89]: Done

File: examples/vision/metric_learning.py (hunk 15 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py]: Done

File: examples/vision/metric_learning.py (hunk 5 above)
Comment (owenvallis) [reply examples/vision/metric_learning.py]: Spoke offline and will add a print statement to make the batch size clear.

**Comment by owenvallis** [comment]
The example depends on tensorflow.similarity 0.14, so I think we need to wait until 0.14 is merged into tf.sim master.

File: examples/vision/metric_learning_tf_similarity.py [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+To make this easy, tf_similarity offers `Samplers()` that enable you to set both
+the number of classes and the minimum number of examples of each class per
+batch.
+
+The train and test datasets will be created using the
+`TFDatasetMultiShotMemorySampler()`. This creates a sampler that loads datasets
+from [tensorflow datasets](https://www.tensorflow.org/datasets) and yields
+batches containing a target number of classes and a target number of examples
+per class. Additionally, we can restrict the sampler to only yield the subset of
+classes defined in class_list, enabling us to train on a subset of the classes
```
Comment (fchollet) [inline examples/vision/metric_learning_tf_similarity.py:95]: Use backticks around code keywords (we'll do a separate round of copyediting at the end)

File: examples/vision/metric_learning_tf_similarity.py [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+
+import random
+
+from matplotlib import pyplot as plt
+from mpl_toolkits.axes_grid1 import ImageGrid
+import numpy as np
+
+import tensorflow as tf
+from tensorflow.keras import layers
+from tensorflow.keras.optimizers import Adam
```
[the remark this answers is not in the document]
Comment (owenvallis) [reply examples/vision/metric_learning_tf_similarity.py:49]: Done. I changed it to `from tensorflow import keras` and also updated the layers.

File: examples/vision/metric_learning_tf_similarity.py [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+)  # allows to interactively explore the samples in 2D space
+from tensorflow_similarity.visualization import (
+    viz_neigbors_imgs,
+)  # neighbors visualization
+
+
+tfsim.utils.tf_cap_memory()
+
+print("TensorFlow:", tf.__version__)
+print("TensorFlow Similarity", tfsim.__version__)
```
[the remark this answers is not in the document]
Comment (owenvallis) [reply examples/vision/metric_learning_tf_similarity.py:74]: Done

File: examples/vision/metric_learning_tf_similarity.py [hunk 19]
```diff
@@ -0,0 +1,432 @@
...
+select a block of samples.
+
+Alternatively, we can use the `generate_batch()` method to yield a batch. This
+can allow us to check that a batch contains the expected number of classes and
+examples per class.
+"""
+
+num_cols = num_rows = 5
+# Get the first 25 examples.
+X, y = train_ds.get_slice(begin=0, size=num_cols * num_rows)
```
[the remark this answers is not in the document]
Comment (owenvallis) [reply examples/vision/metric_learning_tf_similarity.py:164]: Done.

File: examples/vision/metric_learning_tf_similarity.py [hunk 20]
```diff
@@ -0,0 +1,432 @@
...
+class, from which it computes the loss over the pairwise positive and negative
+distances. Here we are using `MultiSimilarityLoss()`
+([paper](ihttps://arxiv.org/abs/1904.06627)), one of several losses in
+[Tensorflow Similarity](https://github.com/tensorflow/similarity). This loss
+attempts to use all informative pairs in the batch, taking into account the
+self-similarity, positive-similarity, and the negative-similarity.
+"""
+
+epochs = 3
+LR = 0.002
```
[the remark this answers is not in the document]
Comment (owenvallis) [reply examples/vision/metric_learning_tf_similarity.py:225]: Done

File: examples/vision/metric_learning_tf_similarity.py [hunk 21]
```diff
@@ -0,0 +1,432 @@
...
+* evaluating the quality of the embedding space
+
+See [Tensorflow Similarity](https://github.com/tensorflow/similarity) for more
+details.
+"""
+
+embedding_size = 256
+
+inputs = layers.Input((32, 32, 3))
+x = inputs
```
[the remark this answers is not in the document]
Comment (owenvallis) [reply examples/vision/metric_learning_tf_similarity.py:195]: Done.

File: examples/vision/metric_learning_tf_similarity.py (hunk 16 above)
Comment (owenvallis) [reply examples/vision/metric_learning_tf_similarity.py:95]: Done. Went through the rest of the document as well to try catch any other code keywords.
